database/db.py

```diff
@@ -115,50 +71,186 @@ def __init__(self, username: str):
         super().__init__(f'user with username {username!r} was not found.')
 
 
-class FriendRequestNotFoundException(FailureException):
+class FriendRequestNotFoundException(DatabaseFailureException):
     """An exception to be raised if a friend request cannot be found."""
 
     def __init__(self, username, friend_username):
         super().__init__(f'No friend request found from {friend_username} on {username}\'s list.')
 
 
+@dataclass
+class DBTransaction:
+    statements: list[str] = field(default_factory=list)
+    was_committed: bool = False
+    was_rolled_back: bool = False
+
+    def add_statement(self, statement: str) -> None:
+        if self.was_committed:
+            raise TransactionClosedException("Cannot add a statement to a transaction that has been committed")
+
+        if self.was_rolled_back:
+            raise TransactionClosedException("Cannot add a statement to a transaction that has been rolled back")
+
+        self.statements.append(statement)
+
+    @property
+    def is_closed(self):
+        return self.was_committed or self.was_rolled_back
```

Comment:
Shifted things to work by transaction. Each transaction can have multiple queries (statements) but can only be committed or rolled back once, and then the transaction is closed.

Reply:
Also played with `@dataclass` a bit for the first time (I think).